lib/libzfs_core/libzfs_core.c

```diff
@@ -1570,12 +1570,17 @@ wait_common(const char *pool, zpool_wait_activity_t activity, boolean_t use_tag,
 {
 	nvlist_t *args = fnvlist_alloc();
 	nvlist_t *result = NULL;
+	int error;
 
 	fnvlist_add_int32(args, ZPOOL_WAIT_ACTIVITY, activity);
 	if (use_tag)
 		fnvlist_add_uint64(args, ZPOOL_WAIT_TAG, tag);
+again:
+	error = lzc_ioctl(ZFS_IOC_WAIT, pool, args, &result);
 
-	int error = lzc_ioctl(ZFS_IOC_WAIT, pool, args, &result);
+	/* macOS gets interrupted */
+	if (error == EINTR)
+		goto again;
```

Comment:
I wonder if this retry loop would be better placed in `libzfs` rather than `libzfs_core`. My understanding is that `libzfs_core` is designed to be consumable by third-party code, and I could image some code using a signal to intentionally wake up a thread doing an `lzc_wait`. Such code wouldn't want this EINTR suppressed.

Also, out of curiousity, do you know where this EINTR is coming from? If this is a spurious wakeup, I wonder if it would be possible to detect that it is spurious in the kernel and handle it there.

Reply:
I've moved it up one level.

and looking at EINTR, it actually comes from ZFS:
`  1  80768       zfsdev_ioctl_common:return 4`
which is odd, as you can not return anything but 0 on macOS (I store the actual returncode in the ioctl struct). I'll dig a little deeper.

Reply:
Ah ok, `spl_wait_sig()` is opposite on macOS. Fixing this as well.